arangod/Agency/State.cpp

```diff
@@ -295,32 +296,38 @@ index_t State::logNonBlocking(index_t idx, velocypack::Slice const& slice,
     }
   }
 
+  logEmplaceBackNoLock(log_t(idx, term, buf, clientId), leading);
+
+  return _log.back().index;
+}
+
+
+void State::logEmplaceBackNoLock(log_t const& l, bool leading) {
+  
   try {
-    _log.push_back(log_t(idx, term, buf, clientId));  // log to RAM or die
+    _log.emplace_back(l);  // log to RAM or die
   } catch (std::bad_alloc const&) {
     if (leading) {
       LOG_TOPIC(FATAL, Logger::AGENCY)
           << "RAFT leader fails to allocate volatile log entries!";
       FATAL_ERROR_EXIT();
     } else {
       LOG_TOPIC(ERR, Logger::AGENCY)
-          << "RAFT follower fails to allocate volatile log entries!";
-      return 0;
+        << "RAFT follower fails to allocate volatile log entries!";
```

Comment:
You have removed a `return 0` here in the follower case. If the above allocation fails and we are in a follower, then we fall through and put the entry into `_clientIdLookupTable` and our invariant is violated from then on. This would be a bug which could take years to show up and then days to find.

Reply:
true